Fix a selection of `const` related issues with `ranget`.

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Attempting to use `ranget` on const input collections or attempting to use a `const ranget` would previously cause compile errors. This PR fixes these issues.

thomasspriggs committed:
Fix constness on dereferencing a `filter_iteratort`

Dereferencing a `filter_iteratort` from filtering a const collection
with `ranget` would previously introduce a compile error, due to a
`const` to non-const conversion. This commit fixes this, by using the
same return type as the iterator which it wraps around.

This commit includes a test for using range to filter a `const`
collection.

thomasspriggs committed:
Fix constness on dereferencing a `concat_iteratort`

Dereferencing a `concat_iteratort` from concating const collections
with `ranget` would previously introduce a compile error, due to a
`const` to non-const conversion. This commit fixes this, by using the
same return type as the iterator which it wraps around.

This commit includes a test for using range to concat `const`
collections.

thomasspriggs committed:
Add `const` to `begin()` accessor of `ranget`

It was previously impossible to iterate over a `const` range, because
you couldn't call `.begin()` on it. This fixes that issue. Changing this
to `const` does not allow the internals of the range to be mutated, as
it returns a copy of the begin iterator, not a reference to it.

This commit adds `const` to ranges in range unit test to show that
ranges can be made const correct, using this fix.

thomasspriggs committed:
Rename `ranget::value_typet` to `ranget::value_type`

Algorithms which are written expecting a STL container may reference a
`value_type` member, as this is the name of the member in STL
containers. Therefore this needs to be named the same way in `ranget` in
order to be compatible with such algorithms.

This commit includes a unit test of `ranget` having correct the
`value_type` defined.

thomasspriggs committed:
Add unit test on range.map over a const container

Given that `filter` and `concat` didn't work with const input
collections, it seems prudent to test that `map` works with const input
collections as well. There is no corresponding fix to make this test
pass in this PR.

kroening commented:
I would squash this slightly more, e.g., put the test into the commit of the change.

kroening commented:
Also, I'd change the return types of
```
const value_type &operator*() const
const value_type *operator->() const
```
for consistency.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7a5ff904147f020a431e67193346e4108caaca3c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97863971

thomasspriggs committed:
Provide `const` dereference of filter / concat iterators only

Deferencing a `const` iterator which refers to a non-const value should
enable mutation of that value. This is consistent with how iterators in
the STL work. It means that we only need to provide a single const
overload of these operators, not both the const and non-const
variations. So the benefit of this commit is better consistency with the
STL and fewer operator overloads to maintain.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 20cc2dd423d0a5b5b8281cb663524c9664bbdfea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97872239

smowton commented on unit/util/range.cpp:
```diff
@@ -42,16 +50,25 @@ SCENARIO("range tests", "[core][util][range]")
       ++it;
       REQUIRE(it == filtered_range.end());
     }
+    THEN("A const filter iterator can mutate the input collection.")
```
Suggest rephrasing, since "const filter iterator" sounds an awful lot like a `const_iterator`, which this is not

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a89ca18e34f3098ceebb2756122454994306564b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97929087
